### ISSUE TITLE
Unicode conversion issue

### ISSUE THREAD

**Issue by pkese** [pythonnet/pythonnet#1466]
### Environment

-   Pythonnet version: 3.0.0-preview2021-06-04
-   Python version: 3.8.8
-   Operating System: ubuntu 21.04
-   .NET Runtime: 5.0.300

### Details

Unicode characters get lost / mangled during conversion:

```fsharp
scope.Exec("testStr = 'Nom 🍗';");
scope.Exec("""print("python:", testStr);""")
Console.WriteLine($"""dotnet: {scope.Get("testStr").ToString()}""");
```

writes
```
python: Nom 🍗
dotnet: Nom �
```

**Comment by pkese** [comment]
Presumably the problem is surrogate pairs.  
It seems that two-pair unicode sequences get incorrect string length - e.g. in actual letters rather than the amount of unicode characters that are required to properly encode these letters.

```fsharp
scope.Exec("testStr = 'Test 💪...';");
scope.Exec("""print("python:", testStr);""")
Console.WriteLine($"""dotnet: {scope.Get("testStr").ToString()}""");
```
results in
```
python: Test 💪...
dotnet: Test 💪..
```
**^^^ Notice how one dot at the end of string is missing in dotnet version**

**Comment by pkese** [comment]
Analogously, putting two unicode surrogates into string will eat two dots at the end
```
python: Test 💪💪...
dotnet: Test 💪💪.
```

**Comment by pkese** [comment]
It looks this problem goes in both directions:  
- when converting from dotnet, Python expects the string `"foo🐼"` to be 4 characters long, but we somehow create a Python object with Length 5.
- when converting from Python to dotnet, we truncate the string as described above. In my test #1467, it appears that there is some sort of size mismatch and a null pointer is returned somewhere in the call stack.

I found some hints of how Python treats unicode conversion internally https://stackoverflow.com/questions/36098984/python-3-3-c-api-and-utf-8-strings  

Using `PyUnicode_DecodeUTF16` or `PyUnicode_DecodeUTF16Stateful` instead of `PyUnicode_FromKindAndData` would maybe alleviate this problem. But getting these functions imported from .dlls and getting delegates in place is probably beyond my capacity.

**Comment by filmor** [comment]
Thank you for the detailed evaluation, I'll have a look at this this weekend. We indeed treat all strings coming from .NET as UCS2 and that breaks as soon as surrogate pairs are involved.

**Comment by pkese** [comment]
I managed to fix the Python to dotnet path in runtime.cs: https://github.com/pythonnet/pythonnet/pull/1467/commits/1b6c6c059b5adc6b09d3673a1d4034c40902b777

This fixes one side of the problem - i.e. the initial bug report.

However the dotnet -> Python conversion issue still remains.

**Comment by filmor** [comment]
Very nice, I'll work on the other direction. The underlying issue (I think) is that C# `"some str".Length` that we use returns the number of Unicode code points, which may be 16 or 32 bit in UTF-16. The safest solution is likely the one you suggested, to use the UTF16-specific functions of the Python API.

**Comment by filmor** [comment]
Fixed by #1467, thank you very much :)

**Comment by pkese** [comment]
And thank You, @filmor, for fixing all the rest.
